src/Shared/PackageManagers/PackageManagerFactory.cs

```diff
@@ -16,29 +18,30 @@ public static class ProjectManagerFactory
         /// <param name="httpClientFactory">The Http cliend factory for making http clients to make requests with.</param>
         /// <param name="destinationDirectory">The directory to use to store any downloaded packages.</param>
         /// <returns></returns>
-        public static BaseProjectManager CreateBaseProjectManager(IHttpClientFactory httpClientFactory, string destinationDirectory)
+        public static BaseProjectManager CreateBaseProjectManager(IHttpClientFactory httpClientFactory, IManagerProvider<IManagerMetadata>? provider, string destinationDirectory)
         {
-            return new BaseProjectManager(httpClientFactory, destinationDirectory);
+            return new BaseProjectManager(httpClientFactory, provider, destinationDirectory);
         }
 
         /// <summary>
         /// Create a BaseProjectManager.
         /// </summary>
         /// <param name="destinationDirectory">The directory to use to store any downloaded packages.</param>
         /// <returns></returns>
-        public static BaseProjectManager CreateBaseProjectManager(string destinationDirectory)
+        public static BaseProjectManager CreateBaseProjectManager(IManagerProvider<IManagerMetadata>? provider, string destinationDirectory)
         {
-            return new BaseProjectManager(destinationDirectory);
+            return new BaseProjectManager(provider, destinationDirectory);
         }
 
         /// <summary>
         /// Get an appropriate project manager for package given its PackageURL.
         /// </summary>
         /// <param name="purl">The <see cref="PackageURL"/> for the package to create the project manager for.</param>
         /// <param name="httpClientFactory"> The <see cref="IHttpClientFactory"/> for the project manager to use for making Http Clients to make web requests.</param>
+        /// <param name="managerProvider">The <see cref="IManagerProvider{IManagerMetadata}"/> for this manager.</param>
         /// <param name="destinationDirectory">The directory to use to store any downloaded packages.</param>
         /// <returns> BaseProjectManager object </returns>
-        public static BaseProjectManager? CreateProjectManager(PackageURL purl, IHttpClientFactory? httpClientFactory = null, string? destinationDirectory = null)
+        public static BaseProjectManager? CreateProjectManager(PackageURL purl, IHttpClientFactory? httpClientFactory = null, IManagerProvider<IManagerMetadata>? managerProvider = null, string? destinationDirectory = null)
```

Comment:
I don't understand how this pattern where you provide the provider here is going to work. It means you have to know ahead of time which manager your purl associates with, which doesn't fit this API surface - which has always been designed to take an arbitrary purl.

Reply:
So my reasoning behind this, similarly to providing the IHttpClientFactory, is that most of the time a provider will not be provided. Just like most of the time we don't provide an IHttpClientFactory, except for our consumption case where we want to control the factory.
By providing the provider in the constructor, it allows us to provide a mocked implementation of the interface for unit test calls. There could very well be a better way of implementing this, but from my still somewhat limited understanding, we need to provide it here. 
I suppose one alternative is that I could add a method to the ProjectManagers that allows us to override and replace the provider that gets added during the factory construction (not from the parameters, but from the ProviderFactory) and then only use that method in unit tests?
If you know of any good design patterns for what I'm trying to accomplish please let me know. It really is only there for the unit test mocking, there shouldn't be a reason from any consuming level side for the consumer to want to override the provider like they might want to with the IHttpClientFactory.

Reply:
What about this:

Add an `IManagerProviderFactory` which takes a purl and an `IHttpClientFactory` and produces the correct `IManagerProvider`.
Each `BaseProjectManager` should have an associated `IManagerProvider`.
Instead of passing an `IHttpClientFactory` pass the `IManagerProviderFactory`.
Optional: the `IManagerProvider` interface should expose the `IHttpClientFactory` that was provided to its factory when it was constructed as well.
When passed a null `IManagerProviderFactory` instantiate the default ones -  (keep same default behavior).
When running unit tests Mock the `IManagerProviderFactory` to produce `IManagerProvider` mocks.

Reply:
This keeps the same flexibility as well to override the `IHttpClientFactory` as before without worrying about writing the `IManagerProvider` as the factory can pass in your `IHttpClientFactory` if you desire.

Reply:
It looks like you may have started a bit with that on the `ProviderFactory.cs`. I think you should just fully commit to the `IProvider` pattern and replace the `HttpClientFactory` parameters entirely (on the managers), with that instead being (optionally) passed to the factory.

Reply:
Just made a huge refactor that did all of that. I think it accomplishes what you were thinking of.